### PR TITLE
Update BenchmarksGame benchmarks to latest

### PR DESCRIPTION
**Pull request by JosephTremoulet**
Previously, for each benchmark, we had a copy of what was the fastest serialized implementation (since the mechanics of tracking these and spotting regressions/improvements through noise are easier than parallel ones).  Update them all to reflect current standings, and to include the best overall for each as well as the best serial implementation for each.  Test names are suffixed to match the indices used to distinguish them on the Benchmarks Game website.

### CODE REVIEW

**Comment by JosephTremoulet** [comment]
@dotnet/jit-contrib

**Comment by JosephTremoulet** [comment]
@dotnet/rap-team

**Comment by jorive** [comment]
@dotnet-bot test Windows_NT x64 perf
@dotnet-bot test Windows_NT x86 perf
@dotnet-bot test linux perf flow

**Comment by JosephTremoulet** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test ([failure](https://ci.dot.net/job/dotnet_coreclr/job/master/job/checked_ubuntu_flow_prtest/10815/) was an I/O error)

File: tests/src/JIT/Performance/CodeQuality/BenchmarksGame/reverse-complement/harness-helpers.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// Helper functionality to locate inputs and find outputs for
+// k-nucleotide benchmark in CoreCLR test harness
```
Comment (AndyAyersMS) [inline tests/src/JIT/Performance/CodeQuality/BenchmarksGame/reverse-complement/harness-helpers.cs]: Typo: "for reverse-complement"

File: tests/src/JIT/Performance/CodeQuality/BenchmarksGame/reverse-complement/reverse-complement-6.cs [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+using System.Threading;
+using Microsoft.Xunit.Performance;
+using Xunit;
+
+[assembly: OptimizeForBenchmarks]
+
+namespace BenchmarksGame
+{
+    class RevCompSequence { public List<byte[]> Pages; public int StartHeader, EndExclusive; public Thread ReverseThread; }
+
```
Comment (AndyAyersMS) [inline tests/src/JIT/Performance/CodeQuality/BenchmarksGame/reverse-complement/reverse-complement-6.cs]: Formatting looks strange here.

File: tests/src/JIT/Performance/CodeQuality/BenchmarksGame/regex-redux/harness-helpers.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// Helper functionality to locate inputs and find outputs for
+// k-nucleotide benchmark in CoreCLR test harness
```
Comment (AndyAyersMS) [inline tests/src/JIT/Performance/CodeQuality/BenchmarksGame/regex-redux/harness-helpers.cs]: Typo "For regex-redux"

File: tests/src/JIT/Performance/CodeQuality/BenchmarksGame/fasta/fasta-1.cs [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+
+            public Frequency(char c, double p)
+            {
+                this.c = (byte)c;
+                this.p = (p * IM);
+            }
+        }
+
+        static Frequency[] IUB = {
+    new Frequency ('a', 0.27)
```
Comment (AndyAyersMS) [inline tests/src/JIT/Performance/CodeQuality/BenchmarksGame/fasta/fasta-1.cs]: Formatting looks odd here

File: tests/src/JIT/Performance/CodeQuality/BenchmarksGame/fasta/fasta-1.cs [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+        ,new Frequency ('N', 0.02)
+        ,new Frequency ('R', 0.02)
+        ,new Frequency ('S', 0.02)
+        ,new Frequency ('V', 0.02)
+        ,new Frequency ('W', 0.02)
+        ,new Frequency ('Y', 0.02)
+};
+
+        static Frequency[] HomoSapiens = {
+    new Frequency ('a', 0.3029549426680)
```
Comment (AndyAyersMS) [inline tests/src/JIT/Performance/CodeQuality/BenchmarksGame/fasta/fasta-1.cs]: Formatting looks odd here

File: tests/src/JIT/Performance/CodeQuality/BenchmarksGame/reverse-complement/reverse-complement-6.cs [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+using System.Security.Cryptography;
+using System.Threading;
+using Microsoft.Xunit.Performance;
+using Xunit;
+
+[assembly: OptimizeForBenchmarks]
+
+namespace BenchmarksGame
+{
+    class RevCompSequence { public List<byte[]> Pages; public int StartHeader, EndExclusive; public Thread ReverseThread; }
```
Comment (jorive) [inline tests/src/JIT/Performance/CodeQuality/BenchmarksGame/reverse-complement/reverse-complement-6.cs]: >class RevCompSequence { public List<byte[]> Pages; public int StartHeader, EndExclusive; public Thread ReverseThread; } [](start = 4, length = 119)

On vs or vscode, Ctrl+A, Ctrl+K+F

**Comment by JosephTremoulet** [comment]
Updated with formatting/typo updates and links to BenchmarksGame CVS repo w/ revision numbers.
Also added some diagnostic output to help determine why the OS X run is failing to find the input files.

**Comment by JosephTremoulet** [comment]
> Also have you looked at per-iteration times to see if they're reasonable?

Yes, I set `InnerIterationCount` for each benchmark to make the reported duration ~1000ms on my dev box.

**Comment by ViktorHofer** [comment]
I didn't know that we store the benchmarks from BenchmarkGame in coreclr. Is that to improve our internal benchmarks and/or to test possible improvements to later submit to http://benchmarksgame.alioth.debian.org/?

**Comment by JosephTremoulet** [comment]
@ViktorHofer it's so that our internal benchmarking will track them, and so we can use that to track the effect of compiler/library/runtime changes on them.  They're not intended to diverge from the sources on the official site, except where needed to run in our benchmarking system.  They might make a convenient place to test out possible source improvements, but we wouldn't check in such a change; instead, after it was submitted to the official site and became the new best variant, we'd pull that new variant down and retire the old one from our suite.

**Comment by ViktorHofer** [comment]
I see, that makes sense! I'm currently looking at those benchmarks and trying to find optimization paths, that's why I asked.

Is there a special reason why you number them e.g. the regex-redux with 1 and 5? why not just name them serial and parallel?

**Comment by JosephTremoulet** [comment]
> Is there a special reason why you number them e.g. the regex-redux with 1 and 5? why not just name them serial and parallel?

From the perspective of using these to track the performance impact of compiler/library/runtime changes, looking at historical trendlines it's most helpful if they represent fixed sources.  If we went with the serial/parallel naming scheme, then when we update to a new fastest variant, there'd be a jump in the historical trendline to stumble over.  So I went the route of using the numbers for the names and indicating serial/parallel in comments.  Less importantly, this also meant not having to figure out what to name a benchmark when the fastest implementation is a serial one.

**Comment by ViktorHofer** [comment]
For historical trendlines it definitely makes sense. I was just thinking about keeping the names to reduce noise when the benchmarks change. But the historical argument is definitely stronger 👍

**Comment by JosephTremoulet** [comment]
OS X passing with latest update, so I'll kick off those perf tests again:

@dotnet-bot test Windows_NT x64 perf
@dotnet-bot test Windows_NT x86 perf
@dotnet-bot test linux perf flow

**Comment by JosephTremoulet** [comment]
The `Windows_NT arm Cross Checked Build and Test` failure is happening because I've removed the old tests but the test script is still trying to run them.  The `Tests.lst` file that lists the old tests has a big comment at the top saying that it's auto-generated...

 1. What do I need to to do re-generate the file?
 2. Why doesn't it get re-generated as part of running the CI?
 3. Failing that, why doesn't the comment telling me the file is auto-generated tell me how to re-generate it?

not sure to whom I should be directing these questions... @jashook?

**Comment by BruceForstall** [comment]
@JosephTremoulet It's auto-generated by @jashook, manually. You can (temporarily) change the tests.lst file to cause the removed tests not to run by changing the Categories to:
```
Categories=EXPECTED_FAIL
```

**Comment by JosephTremoulet** [comment]
Thanks, @BruceForstall.  Updated per your suggestion.

@dotnet-bot test Windows_NT x64 perf
@dotnet-bot test Windows_NT x86 perf
@dotnet-bot test linux perf flow

**Comment by JosephTremoulet** [comment]
@AndyAyersMS, good with updates?

**Comment by AndyAyersMS** [comment]
Yep, thanks for the changes.

**Pull request merged**